cmd/mailserver-canary/main.go

```diff
@@ -0,0 +1,387 @@
+	if err != nil {
+		writeExitMessageFunc = func() { logger.Error("Failed to join public chat", "error", err) }
+		return
+	}
+
+	// watch for envelopes to be available in filters in the client
+	envelopeAvailableWatcher := make(chan whisper.EnvelopeEvent, 1024)
+	sub := clientWhisperService.SubscribeEnvelopeEvents(envelopeAvailableWatcher)
+	defer sub.Unsubscribe()
+
+	// watch for mailserver responses in the client
+	mailServerResponseWatcher := make(chan whisper.EnvelopeEvent, 1024)
+	sub = clientWhisperService.SubscribeEnvelopeEvents(mailServerResponseWatcher)
+	defer sub.Unsubscribe()
+
+	// request messages from mailbox
+	requestID, err := requestHistoricMessages(
+		clientRPCClient,
+		mailboxPeerStr,
+		mailServerKeyID,
+		topic.String(),
+		clientWhisperService.GetCurrentTime().Add(-time.Duration(*period)*time.Second),
+		time.Unix(0, 0),
+		1, "")
+	if err != nil {
+		exitCode = 2
+		writeExitMessageFunc = func() { logger.Error("Error requesting historic messages from mailserver", "error", err) }
+		return
+	}
+
+	// wait for mailserver response
+	resp, err := waitForMailServerResponse(mailServerResponseWatcher, requestID, 10*time.Second)
+	if err != nil {
+		exitCode = 3
+		writeExitMessageFunc = func() { logger.Error("Error waiting for mailserver response", "error", err) }
+		return
+	}
+
+	// wait for last envelope sent by the mailserver to be available for filters
+	err = waitForEnvelopeEvents(envelopeAvailableWatcher, []string{resp.LastEnvelopeHash.String()}, whisper.EventEnvelopeAvailable)
+	if err != nil {
+		exitCode = 4
+		writeExitMessageFunc = func() { logger.Error("Error waiting for envelopes to be available to client filter", "error", err) }
+		return
+	}
+
+	exitCode = 0
+}
+
+// makeNodeConfig parses incoming CLI options and returns node configuration object
+func makeNodeConfig() (*params.NodeConfig, error) {
+	err := error(nil)
+
+	workDir := ""
+	if path.IsAbs(*homePath) {
+		workDir = *homePath
+	} else {
+		workDir, err = filepath.Abs(filepath.Dir(os.Args[0]))
+		if err == nil {
+			workDir = path.Join(workDir, *homePath)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	nodeConfig, err := params.NewNodeConfig(path.Join(workDir, ".ethereum"), "", uint64(params.RopstenNetworkID))
+	if err != nil {
+		return nil, err
+	}
+
+	if *logLevel != "" {
+		nodeConfig.LogLevel = *logLevel
+		nodeConfig.LogEnabled = true
+	}
+
+	if *logFile != "" {
+		nodeConfig.LogFile = *logFile
+	}
+
+	nodeConfig.NoDiscovery = true
+
+	return whisperConfig(nodeConfig)
+}
+
+// whisperConfig creates node configuration object from flags
+func whisperConfig(nodeConfig *params.NodeConfig) (*params.NodeConfig, error) {
+	whisperConfig := nodeConfig.WhisperConfig
+	whisperConfig.Enabled = true
+	whisperConfig.LightClient = true
+	whisperConfig.MinimumPoW = *minPow
+	whisperConfig.TTL = *ttl
+	whisperConfig.EnableNTPSync = true
+
+	return nodeConfig, nil
+}
+
+func startClientNode() (*api.StatusBackend, func(), error) {
```

Comment:
Not sure I follow. The second argument is needed since it allows writing out an error message in case of failure. To be honest, I was torn about extracting this to a function, since it made the function more complex at the expense of a simplified `main()`.

Reply:
It's used like this:
```go
clientBackend, stopFunc, err := startClientNode()
defer stopFunc()
```

while the same can be done like this:
```go
clientBackend, err := startClientNode()
if err != nil {
  log.Crit(err)
  os.Exit(1)
}
defer func() { clientBackend.StopNode() }
```

Reply:
The reasoning is the same as above, that wouldn't guarantee that the log message would be displayed last.

Reply:
The initial approach was probably being too conservative in order to guarantee that no matter what happened, the message would be displayed last. I agree that we can be less generic and make the code a bit more readable (see latest commit).